### PR TITLE
Python 3 support

### PR DESCRIPTION
**Pull request by maratkomarov**
This PR adds Python 2 / 3 support

### CODE REVIEW

**Comment by coderanger** [comment]
As it stands this has a ton of semantics changes on Python 2. It would need massive cleanup to be viable. A good start though!

File: chef/api.py [hunk 1]
```diff
@@ -80,7 +87,7 @@ def __init__(self, url, key, client, version='0.10.8', headers={}):
         self.key = key
         self.client = client
         self.version = version
-        self.headers = dict((k.lower(), v) for k, v in headers.iteritems())
+        self.headers = dict((k.lower(), v) for k, v in headers.items())
```
Comment (coderanger) [inline chef/api.py]: This would have to use `six.iteritems` to match semantics on Python 2. This is a common issue below so not pointing out each instance.

File: chef/api.py [hunk 2]
```diff
@@ -99,18 +106,19 @@ def from_config_file(cls, path):
...
-                continue # Not a simple key/value, we can't parse it anyway
+                continue  # Not a simple key/value, we can't parse it anyway
             key, value = parts
             md = cls.ruby_string_re.search(value)
             if md:
                 value = md.group(2)
             else:
                 # Not a string, don't even try
-                log.debug('Value for %s does not look like a string: %s'%(key, value))
+                log.debug('Value for %s does not look like a string: %s' % (key, value))
```
Comment (coderanger) [inline chef/api.py]: Should probably be swapped to use `.format`

File: chef/auth.py [hunk 3]
```diff
@@ -71,10 +81,10 @@ def sign_request(key, http_method, path, body, host, timestamp, user_id):
...
         'x-ops-content-hash': hashed_body,
     }
-    
+
     # Create RSA signature
     req = canonical_request(http_method, path, hashed_body, timestamp, user_id)
     sig = _ruby_b64encode(key.private_encrypt(req))
     for i, line in enumerate(sig):
-        headers['x-ops-authorization-%s'%(i+1)] = line
+        headers['x-ops-authorization-%s' % (i + 1)] = line
```
Comment (coderanger) [inline chef/auth.py]: Please undo all these formatting-only changes. It makes the rest of the diff hard to read.

File: chef/fabric.py [hunk 4]
```diff
@@ -38,7 +39,7 @@ def __init__(self, query, api, hostname_attr, environment=None):
         self.query = query
         self.api = api
         self.hostname_attr = hostname_attr
-        if isinstance(self.hostname_attr, basestring):
+        if isinstance(self.hostname_attr, str):
```
Comment (coderanger) [inline chef/fabric.py]: This is a semantic change, should be `six.string_types`.

File: chef/fabric.py [hunk 5]
```diff
@@ -51,7 +52,7 @@ def __call__(self):
             query += ' AND chef_environment:%s' % environment
         for row in Search('node', query, api=self.api):
             if row:
-                if callable(self.hostname_attr):
```
Comment (coderanger) [inline chef/fabric.py]: `callable` still exists in in 3.2+ and is preferable in this situation.

File: chef/rsa.py [hunk 6]
```diff
@@ -8,28 +9,30 @@
 else:
     _eay = CDLL('libcrypto.so')
 
-#unsigned long ERR_get_error(void);
+# unsigned long ERR_get_error(void);
```
Comment (coderanger) [inline chef/rsa.py]: Mentioned above, but please revert what I assume was running flake8 on this. That can be its own PR.

File: chef/rsa.py [hunk 7]
```diff
@@ -141,23 +152,22 @@ def __init__(self, fp=None):
         self.public = False
         if not fp:
             return
-        if isinstance(fp, basestring):
-            if fp.startswith('-----'):
-                # PEM formatted text
-                self.raw = fp
-            else:
-                self.raw = open(fp, 'rb').read()
+        if isinstance(fp, six.binary_type) and fp.startswith('-----'.encode()):
```
Comment (coderanger) [inline chef/rsa.py]: Should probably be `b'-----'` if we are requiring it to be a bytes type.

File: chef/tests/__init__.py [hunk 8]
```diff
@@ -50,12 +50,12 @@ def tearDown(self):
...
+            except ChefError as e:
+                print(e)
                 # Continue running
 
     def register(self, obj):
         self.objects.append(obj)
 
     def random(self, length=8, alphabet='0123456789abcdef'):
-        return ''.join(random.choice(alphabet) for _ in xrange(length))
+        return ''.join(random.choice(alphabet) for _ in range(length))
```
Comment (coderanger) [inline chef/tests/__init__.py]: Use the six helper.

File: chef/utils/json.py [hunk 9]
```diff
@@ -1,12 +1,13 @@
...
 import types
+import collections
 try:
     import json
 except ImportError:
     import simplejson as json
 
 def maybe_call(x):
-    if callable(x):
+    if isinstance(x, collections.Callable):
```
Comment (coderanger) [inline chef/utils/json.py]: As above, callable exists in 3.2+.

**Comment by maratkomarov** [comment]
@coderanger we've spread `six` usage and removed formatting changes, unrelated to the topic. Take a look.

**Comment by coderanger** [comment]
Awesome, thanks for doing this!
